inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java

```diff
@@ -42,22 +45,94 @@
     private String stringContent;
 
     /**
-     * Contains position mapping for a character between PDFExtract string
+     * Contains position mapping for characters between PDFExtract string
      * including and excluding Draw Operations.
      */
     private BidiMap<Integer, Integer> stringPositionMap;
 
+    private Map<String, String> ligatures;
+
+    private String ligaturelessContent;
+
+    /**
+     * Contains position mapping for characters between ligaturelessContent and stringContent
+     */
+    private BidiMap<Integer, Integer> ligatureStringPositionMap;
+
     /**
      * Map of line numbers and lines contained in a PDFExtract file.
      */
     private Map<Integer, PdfExtractLine> extractLines;
 
-    public PdfExtractFile(String aPdftxt)
+
+    public PdfExtractFile(String aPdftxt, Map<String, String> aLigatures)
     {
-        setPdftxt(aPdftxt);
+        initializeStringContent(aPdftxt);
+        initializeLigaturelessContent(aLigatures);
+    }
+
+    private void initializeLigaturelessContent(Map<String, String> aLigatures)
+    {
+        ligatures = aLigatures;
+        ligaturelessContent = stringContent;
+        ligatureStringPositionMap = new DualHashBidiMap<>();
+
+        // build Aho-Corasick Trie to search for ligature occurences and replace them
+        Trie.TrieBuilder trieBuilder = Trie.builder();
+        ligatures.keySet().forEach(key -> trieBuilder.addKeyword(key));
+        Trie trie = trieBuilder.build();
+        Collection<Emit> emits = trie.parseText(ligaturelessContent);
+        Map<Integer, Emit> occurrences = new HashMap<>();
+        for (Emit emit : emits) {
+            occurrences.put(emit.getStart(), emit);
+        }
+
+        int stringIndex = 0;
+        int ligatureIndex = 0;
+        StringBuilder sb = new StringBuilder();
+        // iterate over the text and create a new string containing replaced ligatures
+        // also create a new map that maps from new ligature string to normal string content
+        while (stringIndex < ligaturelessContent.length()) {
+            char c = ligaturelessContent.charAt(stringIndex);
+            if (occurrences.containsKey(stringIndex)) {
+                // start of a ligature was found
+                Emit emit = occurrences.get(stringIndex);
+                String replacement = ligatures.get(emit.getKeyword());
+                int keyLen = emit.getKeyword().length();
+                int replacementLen = replacement.length();
+
+                // iterate over chars of ligature string and its replacement, create proper mapping
+                ligatureStringPositionMap.put(ligatureIndex, stringIndex);
+                sb.append(replacement.charAt(0));
```

Comment:
Why do you only get the first character here? The substitution map contains some substitutions which go from 3 characters in length to 2 characters

Reply:
Remaining characters will be added later in the for loop.
I get the first one because I assume that we do have no empty replacements (e.g. "sample" to "") and it helps with indices and if coniditions in the loop.